katdal/categorical.py

```diff
@@ -591,69 +654,92 @@ def sensor_to_categorical(sensor_timestamps, sensor_values, dump_midtimes, dump_
+    first_proper_event = events.searchsorted(-1, side='right')
+    # Shift the final prior event (if any) to the start of the first dump
+    if first_proper_event > 0:
+        first_proper_event -= 1
+        events[first_proper_event] = 0
+    one_past_last_event = events.searchsorted(num_dumps)
+    within_dumps = slice(first_proper_event, one_past_last_event)
+    sensor_values = sensor_values[within_dumps]
+    events = np.r_[events[within_dumps], num_dumps]
     # Apply optional transform to sensor values
     if transform is not None:
         if wrapped_values:
-            unwrapped_transform = transform
-            transform = lambda value: ComparableArrayWrapper(unwrapped_transform(value.unwrapped))  # noqa: E731
+            orig_transform = transform
+            def transform(value):   # noqa: E301
+                """Unwrap wrapped value, transform and rewrap."""
+                return ComparableArrayWrapper(orig_transform(value.unwrapped))
         sensor_values = np.array([transform(y) for y in sensor_values])
-    # Discard sensor events that do not change the (transformed) sensor value (i.e. that repeat the previous value)
+    # Force first dump to have valid sensor value
+    # (insert initial value or let the first proper value apply from the start)
+    if events[0] != 0 and initial_value is not None:
+        if wrapped_values:
+            initial_value = ComparableArrayWrapper(initial_value)
+        sensor_values = np.r_[[initial_value], sensor_values]
+        events = np.r_[0, events]
+    events[0] = 0
+    # Clean up dump->event mapping, taking into account greedy values
+    greedy_values = () if greedy_values is None else greedy_values
+    greedy = [value in greedy_values for value in sensor_values]
+    cleaned_up = [event for event in _single_event_per_dump(events, greedy)]
+    sensor_values = sensor_values[cleaned_up]
+    events = np.r_[events[cleaned_up], num_dumps]
+    # Discard sensor events that do not change the (transformed) sensor value
+    # (i.e. that repeat the previous value)
     if not allow_repeats:
-        changes = [n for n in range(len(sensor_values)) if (n == 0) or (sensor_values[n] != sensor_values[n - 1])]
-        sensor_values, events = sensor_values[changes], np.r_[events[changes], len(dump_midtimes)]
-    # Extend segments with "greedy" values to include first dump of next segment where sensor value is changing
-    if greedy_values is not None:
-        greedy_to_nongreedy = [n for n in range(1, len(sensor_values))
-                               if sensor_values[n - 1] in greedy_values and sensor_values[n] not in greedy_values]
-        events[greedy_to_nongreedy] += 1
+        changes_value = [n for n in range(len(sensor_values)) if (n == 0) or
+                         (sensor_values[n] != sensor_values[n - 1])]
+        sensor_values = sensor_values[changes_value]
+        events = np.r_[events[changes_value], num_dumps]
```

Comment:
I was actually thinking of also combining the values + events filtering into a single function as it happens thrice, only with different selections. But your suggestion seems better, done!